### DISCUSSION TITLE
Themes and OnAfterRenderAsync

### DISCUSSION THREAD

**Discussion by leigh-pointer** [oqtane/oqtane.framework#4171] [General]
I am having an issue trying to execute s JSRuntime.InvokeVoidAsync call in the OnAfterRenderAsync method. I am checking if (firstRender) 

ERROR
```
Microsoft.JSInterop.JSException: 'Could not find 'Startup.init' ('Startup' was undefined).
Error: Could not find 'Startup.init' ('Startup' was undefined).
    at http://localhost:44357/_framework/blazor.web.js:1:734
    at Array.forEach (<anonymous>)
    at l.findFunction (http://localhost:44357/_framework/blazor.web.js:1:702)
    at b (http://localhost:44357/_framework/blazor.web.js:1:5445)
    at http://localhost:44357/_framework/blazor.web.js:1:3238
    at new Promise (<anonymous>)
    at y.beginInvokeJSFromDotNet (http://localhost:44357/_framework/blazor.web.js:1:3201)
    at gn._invokeClientMethod (http://localhost:44357/_framework/blazor.web.js:1:62841)
    at gn._processIncomingData (http://localhost:44357/_framework/blazor.web.js:1:60316)
    at connection.onreceive (http://localhost:44357/_framework/blazor.web.js:1:53957)'
```

Am I missing some voodoo magic ?

**Comment by sbwalker** [comment]
@leigh-pointer can you share your OnAfterRenderAsync code? JS Interop is painful as there is so much potential for race conditions

**Reply by leigh-pointer** [reply]
Moving only  this code to a new component and placing to on the Theme control solved the problem.  
But adding delay to the Theme didn't work.

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        try
        {
            if (firstRender)
            {
                await Task.Delay(100);
                await JSRuntime.InvokeVoidAsync("Startup.init", new object[] { new { } });

            }

            await JSRuntime.InvokeVoidAsync("Startup.refresh");

        }
        catch 
        {
            await base.Log(PageState.Alias, LogLevel.Error, "InitJS.OnAfterRenderAsync", ex, "InvokeVoidAsync Script");
        }
    }

**Comment by sbwalker** [comment]
@leigh-pointer how are you loading the JavaScript which is being used by the Theme component. When using Interactive Blazor and JS Interop you need to coordinate the loading of the JavaScript with the execution of the JavaScript. Usually this means that you need to register the JavaScript in the Resources collection within the Theme component and you need to call b await base.OnAfterRenderAsync(firstRender); as the first step in your Theme component's OnAfterRenderAsync() event. Calling  await base.OnAfterRenderAsync(firstRender); is what actually loads the JavaScript into the page. Once it is loaded then you can execute one of its methods. However it can get very tricky if you have StateHasChanged() logic elsewhere in your component as the life cycle events can fire out of order.

**Reply by leigh-pointer** [reply]
doh!! I wasn't calling  `base.OnAfterRenderAsync(firstRender); `
Works ok now, which is better now it is using the framework.
